Source/WebGPU/WGSL/TypeCheck.cpp

```diff
@@ -0,0 +1,321 @@
+    Type* functionType = nullptr;
+    ContextProvider::introduceVariable(function.name(), functionType);
+}
+
+// Statements
+void TypeChecker::visit(AST::Statement& statement)
+{
+    UnificationContext statementContext(this);
+    AST::Visitor::visit(statement);
+}
+
+void TypeChecker::visit(AST::AssignmentStatement& statement)
+{
+    auto* lhs = infer(statement.lhs());
+    auto* rhs = infer(statement.rhs());
+    unify(lhs, rhs);
+}
+
+void TypeChecker::visit(AST::ReturnStatement& statement)
+{
+    // FIXME: handle functions that return void
+    auto* type = infer(*statement.maybeExpression());
+
+    // FIXME: unify type with the curent function's return type
+    UNUSED_PARAM(type);
+}
+
+// Expressions
+void TypeChecker::visit(AST::Expression&)
+{
+    // FIXME: remove this function once we start allocating types
+    inferred(&m_type);
+}
+
+void TypeChecker::visit(AST::FieldAccessExpression& access)
+{
+    auto* structType = infer(access.base());
+    // FIXME: implement member lookup once we have a struct type
+    UNUSED_PARAM(structType);
+
+    Type* fieldType = nullptr;
+    inferred(fieldType);
+}
+
+void TypeChecker::visit(AST::IndexAccessExpression& access)
+{
+    // FIXME: handle reference index access
+    auto* arrayType = infer(access.base());
+    auto* index = infer(access.index());
+
+    // FIXME: unify index with UnionType(i32, u32)
+    UNUSED_PARAM(index);
+
+    // FIXME: set the inferred type to the array's member type
+    UNUSED_PARAM(arrayType);
+}
+
+void TypeChecker::visit(AST::BinaryExpression& binary)
+{
+    // FIXME: this needs to resolve overloads, not just unify both types
+    auto* result = variable();
+    auto* leftType = infer(binary.leftExpression());
+    auto* rightType = infer(binary.rightExpression());
+    unify(result, leftType);
+    unify(result, rightType);
+    inferred(result);
+}
+
+void TypeChecker::visit(AST::IdentifierExpression& identifier)
+{
+    auto* const* type = ContextProvider::readVariable(identifier.identifier());
+    // FIXME: this should be unconditional
+    ASSERT(type);
+    if (type)
+        inferred(*type);
+}
+
+// Types
+void TypeChecker::visit(AST::TypeName&)
+{
+    // FIXME: remove this function once we start allocating types
+    inferred(&m_type);
+}
+
+// Private helpers
+Type* TypeChecker::infer(AST::Expression& expression)
+{
+    ASSERT(!m_inferredType);
+    // FIXME: this should call the base class and TypeChecker::visit should assert
+    // that it is never called directly on expressions
+    visit(expression);
+    ASSERT(m_inferredType);
+
+    auto* type = m_inferredType;
+    substitute(type, [&](Type* resolvedType) {
+        // FIXME: store resolved type in the expression
+        UNUSED_PARAM(resolvedType);
+    });
+    m_inferredType = nullptr;
+
+    return type;
+}
+
+Type* TypeChecker::variable()
+{
+    // FIXME: allocate new type variables
+    return nullptr;
+}
+
```

Comment:
I kind of think we should do the Behavior checker too while we're at it, since it composes nicely with the expression type checker (that you've got here).

The Behavior checker is the one that assigns a set of Behaviors to each statement, which correspond to the ways that control flow can exit that statement. From memory, the set of Behaviors is {Passthrough, Continue, Break, Return}. The Behavior of a conditional is the union of the conditions of both of its sides, etc. Using this kind of analysis, we can validate things like "no continue statement can appear outside of a loop" and "if a break appears in a block, it's the last statement in the block" and "if a statement has a non-void return type, all control flow paths must end in a return"